### PR TITLE
NEW Add sort columns to DB index automatically

### LINKED ISSUE #7272: default_sort fields should be added as indexes
Sorting on an indexed field is faster: https://dev.mysql.com/doc/refman/5.7/en/order-by-optimization.html#order-by-index-use. 

Therefore, I would like to suggest that, by default, database fields listed in `default_sort` should be added as indexes. 

Not sure if it would work for `default_sorts` with multiple fields, but, for example, for `SiteTree` we should definitely add an Index on the `Sort` Field.

### PR DESCRIPTION
**Pull request by dhensby**
fixes #7272

Automatically add `default_sort` columns to DB index

### CODE REVIEW

**Comment by dhensby** [comment]
Failures are behat suites that aren't working at the moment

**Comment by tractorcow** [comment]
Can you please add some unit tests?

File: model/DataObject.php [hunk 1]
```diff
@@ -3403,6 +3404,29 @@ public function databaseIndexes() {
 			}
 		}
 
+		if ($sort && is_string($sort)) {
+		    $sort = preg_split('/,(?![^()]*+\\))/', $sort);
```
Comment (tractorcow) [inline model/DataObject.php]: What's the significance of `*+` here?

File: model/DataObject.php [hunk 2]
```diff
@@ -3403,6 +3404,29 @@ public function databaseIndexes() {
...
+
+			foreach ($sort as $value) {
```
Comment (tractorcow) [inline model/DataObject.php]: Please check out ArrayList::parseSortColumn(). This has pre-existing tests, and covers the range of various sort syntaxes a bit more robustly. For instance `$default_sort = '"SiteTree"."Order" DESC';` doesn't seem covered by this code.

Maybe we could extract that helper somewhere that index generation could access it?

Comment (tractorcow) [reply model/DataObject.php]: It looks like that method could also be extended to support multi-columns specifications; There's a chance to improve both ArrayList::sort() as well as index generation. :)

File: model/DataObject.php (hunk 1 above)
Comment (dhensby) [reply model/DataObject.php]: It's copied from `SQLselect::getsortorder` so I don't know :)

File: model/DataObject.php (hunk 2 above)
Comment (dhensby) [reply model/DataObject.php]: I copied code from SQLSelect, happy to copy from somewhere else if it's more robust :)

File: model/DataObject.php (hunk 1 above)
Comment (tractorcow) [reply model/DataObject.php]: Oh! I see... yeah that code itself isn't so easy to understand either. Apologies for implying it was your issue. :)

That code exists as-is in 4.0 too.

I think the SQLSelect::getOrderBy() only needs to return the SQL fragments that'll be sent to the final query; What we want is more heavily parsed, so I guess the ArrayList code is probably closer to our needs in this case?.

Oh well, once you step in a poop you're walking around in it all day aren't you. :)

**Comment by tractorcow** [comment]
Maybe it's time to deprecate string-only sort, and shift to array format like we did with indexes? That turned out really successfully didn't it?

We could shift away from "write sort as ansi-sql strings" to "field lists that the ORM will sql-ify".

**Comment by dhensby** [comment]
> Maybe it's time to deprecate string-only sort, and shift to array format like we did with indexes? That turned out really successfully didn't it?
> 
> We could shift away from "write sort as ansi-sql strings" to "field lists that the ORM will sql-ify".

I'd really like that but I don't think it's realistic for 3 or 4, unfortunately.

I'll look into the other code we have for parsing sort strings and add the `"SiteTree"."Sort"` case to the tests

**Comment by dhensby** [comment]
@tractorcow ok, I've copied (and stripped down) the `parseSortColumn` method from `ArrayList`. It's not really made a big difference to how much code I've had to hand roll because it doesn't handle multiple sorts nor does it strip ANSI quotes.

**Comment by tractorcow** [comment]
> I've had to hand roll because it doesn't handle multiple sorts nor does it strip ANSI quotes.

Just one question before I merge... what happens if you have code like this?

```php
private static $default_sort = '"SiteTree"."Title" ASC';
```

It doesn't cause an error does it?

File: tests/model/DataObjectSchemaGenerationTest.php [hunk 3]
```diff
@@ -212,3 +257,12 @@ class DataObjectSchemaGenerationTest_IndexDO extends DataObjectSchemaGenerationT
...
 	);
 }
+
+class DataObjectSchemaGenerationTest_Sorted extends DataObject implements TestOnly {
+    private static $db = array(
+        'Title' => 'Varchar',
+        'Sort' => 'Int',
+    );
+
+    private static $default_sort = "Sort";
```
Comment (tractorcow) [inline tests/model/DataObjectSchemaGenerationTest.php]: Can you add a test using ansi sql quotes to show it doesn't break those?

**Comment by dhensby** [comment]
I have test cases for ANSI quoted sort and table named sorts.

File: model/DataObject.php [hunk 4]
```diff
@@ -3403,13 +3404,58 @@ public function databaseIndexes() {
...
+						continue;
+					}
+
+					if (substr($column, 0, 1) === '"') {
+						$column = substr($column, 1, strlen($column)-1);
+					}
+
+					if (substr($column, -1, 1) === '"') {
+						$column = substr($column, 0, -1);
+					}
```
Comment (kinglozzer) [inline model/DataObject.php]: Would `trim($column, '"')` have the same effect as these two `substr` blocks?

Comment (dhensby) [reply model/DataObject.php]: hah, that would be a touch more elegant, wouldn't it!

**Pull request merged**